rosetta-source/src/main/rosetta/event-workflow-func.rosetta

```diff
@@ -145,6 +145,38 @@ func Create_AcceptedWorkflowStepFromInstruction: <"Represents the acceptance of
     add acceptedWorkflowStep -> account:
         acceptedWorkflowStep -> businessEvent -> after -> trade -> account distinct
 
+func Create_CancelledWorkflowStepFromInstruction: <"Represents the cancellation of a proposed instruction that results in a workflow step containing a business event, message details, identifiers, event timestamps, parties and accounts. The previous workflow step input must exist to provide workflow lineage.">
+	[creation WorkflowStep]
+	inputs:
+		proposedWorkflowStep WorkflowStep (1..1)
+	output:
+		cancelledWorkflowStep WorkflowStep (1..1)
+
+	condition ProposedEventExists: <"The previous step being cancelled must be a proposed step containing an instruction.">
+		proposedWorkflowStep -> proposedEvent exists
+
+	condition CancelledProposedStep: <"The previous step must be a cancel">
+		proposedWorkflowStep -> action = ActionEnum -> Cancel
+
+	condition RejectedProposedStep: <"The previous step cannot be rejected.">
+		proposedWorkflowStep -> rejected <> True
+
+	set cancelledWorkflowStep -> action: proposedWorkflowStep -> action
+	set cancelledWorkflowStep -> messageInformation: proposedWorkflowStep -> messageInformation
+	add cancelledWorkflowStep -> timestamp: proposedWorkflowStep -> timestamp
+	add cancelledWorkflowStep -> eventIdentifier: proposedWorkflowStep -> eventIdentifier
+	set cancelledWorkflowStep -> previousWorkflowStep: proposedWorkflowStep -> previousWorkflowStep as-key
+
+	set cancelledWorkflowStep -> businessEvent: <"Assign the business event corresponding to the workflow step.">
+		Create_BusinessEvent(
```

Comment:
If we're cancelling a WorkflowStep, and the WorkflowStep has a proposedEvent in it, as confirmed in an earlier condition, why are we now creating a BusinessEvent? Isn't a BusinessEvent only created as a result of an accepted WorkflowStep?

Here you're checking for the existence of a proposedEvent which implies that the workflow has not completed successfully and there should be no BusinessEvent at this point?

If there is a businessEvent on the WorkflowStep already then I can see the logic in including it on the cancelledWorkflowStep, but I can't see why we would want to create a new BusinessEvent for a cancelled WorkflowStep.

Can you give a bit more detail please in case I am missing something.

Reply:
Hi, the proposal is fulfilling the requirement to represent that the previous step was accepted by mistake. We use it as previous state to propose a new instruction, in which the cancellation flag is set and it's executed through the proposed function.

Requirement in DRR was to be able to represent an ERROR on previously submitted states.